samples/extensions/device_fault/device_fault.cpp

```diff
@@ -0,0 +1,628 @@
+    vkBeginCommandBuffer(cmd, &begin_info);
+
+    // First thing is to update the pointer buffer.
+    // We could use a compute shader here if we're doing
+    // GPU-driven rendering for example.
+    update_pointer_buffer(cmd);
+
+    // Arbitrary value between 0 and 1 to create some animation.
+    accumulated_time += 0.2f * delta_time;
+    accumulated_time = glm::fract(accumulated_time);
+
+    // Update VBOs through buffer_device_address.
+    update_meshlets(cmd);
+
+    VkRenderPassBeginInfo render_pass_begin    = vkb::initializers::render_pass_begin_info();
+    render_pass_begin.renderPass               = render_pass;
+    render_pass_begin.renderArea.extent.width  = width;
+    render_pass_begin.renderArea.extent.height = height;
+    render_pass_begin.clearValueCount          = 2;
+    VkClearValue clears[2]                     = {};
+    clears[0].color.float32[0]                 = 0.033f;
+    clears[0].color.float32[1]                 = 0.073f;
+    clears[0].color.float32[2]                 = 0.133f;
+    render_pass_begin.pClearValues             = clears;
+    render_pass_begin.framebuffer              = framebuffers[current_buffer];
+
+    vkCmdBeginRenderPass(cmd, &render_pass_begin, VK_SUBPASS_CONTENTS_INLINE);
+
+    vkCmdBindPipeline(cmd, VK_PIPELINE_BIND_POINT_GRAPHICS, pipelines.bindless_vbo_pipeline);
+    vkCmdSetViewport(cmd, 0, 1, &viewport);
+    vkCmdSetScissor(cmd, 0, 1, &scissor);
+
+    PushVertex push_vertex{};
+
+    // Create an ad-hoc perspective matrix.
+    push_vertex.view_projection =
+            glm::perspective(0.5f * glm::pi<float>(), static_cast<float>(width) / static_cast<float>(height), 1.0f, 100.0f) *
+            glm::lookAt(glm::vec3(0.0f, 0.0f, 5.0f), glm::vec3(0.0f), glm::vec3(0.0f, 1.0f, 0.0f));
+
+    // Push pointer to array of meshlets.
+    // Every instance renders its own meshlet.
+    push_vertex.table = pointer_buffer.gpu_address;
+    vkCmdPushConstants(cmd, pipelines.graphics_pipeline_layout, VK_SHADER_STAGE_VERTEX_BIT, 0, sizeof(push_vertex), &push_vertex);
+    vkCmdBindIndexBuffer(cmd, index_buffer->get_handle(), 0, VK_INDEX_TYPE_UINT16);
+    vkCmdDrawIndexed(cmd, mesh_num_indices, static_cast<uint32_t>(test_buffers.size()), 0, 0, 0);
+
+    draw_ui(cmd);
+
+    vkCmdEndRenderPass(cmd);
+
+    VK_CHECK(vkEndCommandBuffer(cmd));
+    submit_info.commandBufferCount = 1;
+    submit_info.pCommandBuffers    = &draw_cmd_buffers[current_buffer];
+
+    VK_CHECK(vkQueueSubmit(queue, 1, &submit_info, wait_fences[current_buffer]));
+
+    // A bit of a hack. This is usually seated in ApiVulkanSample::submit_frame(), but that throws immediately if the device enters an error state.
+    // So we incorrectly call wait_idle here, so we can get the GPU in error state, and we can query it for device_fault before an exception is thrown.
+    VkResult error = get_device().get_queue_by_present(0).wait_idle();
+    check_device_fault(error);
+    ApiVulkanSample::submit_frame();
```

Comment:
Maybe, instead of this slightly hacky approach:
```
	try
	{
		ApiVulkanSample::submit_frame();
	}
	catch (std::exception const &e)
	{
		vk::DeviceLostError const *device_lost_error = reinterpret_cast<vk::DeviceLostError const *>(&e);
		if (device_lost_error)
		{
			check_device_fault(VK_ERROR_DEVICE_LOST);
		}
	}
```